### PR TITLE
feat(manager/argocd): Add support for Kustomize image overrides

### PR DESCRIPTION
**Pull request by j-be**
## Changes

This PR aims to add support for updating Kustomize image overrides in ArgoCD Application resources (see https://argo-cd.readthedocs.io/en/stable/user-guide/kustomize/ for details).

## Context

See https://github.com/renovatebot/renovate/discussions/23848 for details.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository (https://github.com/j-be/renovate/tree/feat/manager-argocd/add-support-for-kustomize-images)

### CODE REVIEW

File: lib/modules/manager/argocd/__fixtures__/validApplication.yml [hunk 1]
```diff
@@ -129,3 +129,14 @@ spec:
         valueFiles:
           - $foo/values.yaml
 
+---
```
Comment (viceice) [inline lib/modules/manager/argocd/__fixtures__/validApplication.yml:132]: can you add it to an existing application spec?

File: lib/modules/manager/argocd/extract.ts [hunk 2]
```diff
@@ -18,6 +19,8 @@ import {
 } from './schema';
 import { fileTestRegex } from './util';
 
+const kustomizeImageRe = regEx(/(.+)=(.+):(.+)/);
```
Comment (viceice) [inline lib/modules/manager/argocd/extract.ts:22]: this is wrong. it requires a tag and doesn't support digests. tag and digest should be optional 

you should use a similar approach like we already use in kustomize

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/argocd/__fixtures__/validApplication.yml (hunk 1 above)
Comment (j-be) [reply lib/modules/manager/argocd/__fixtures__/validApplication.yml]: Done.

File: lib/modules/manager/argocd/extract.ts [hunk 3]
```diff
@@ -88,15 +107,31 @@ function processAppSpec(
...
+    return null;
+  }
+
+  const parts = kustomizeImageRe.exec(image);
+  if (parts?.length !== 3) {
+    return null;
+  }
+
+  return {
+    ...splitImageParts(parts[2]),
```
[the remark this answers is not in the document]
Comment (j-be) [reply lib/modules/manager/argocd/extract.ts]: Changed.

Not gonna lie: I'm having a hard time reading and navigating dockerfile/extract.

File: lib/modules/manager/argocd/extract.ts [hunk 4]
```diff
@@ -88,15 +107,31 @@ function processAppSpec(
...
   }
 
-  return deps.filter(is.truthy);
+  return deps;
+}
+
+function processKustomizeImage(image: string): PackageDependency | null {
+  if (!kustomizeImageRe.test(image)) {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (j-be) [reply lib/modules/manager/argocd/extract.ts]: Removed

File: lib/modules/manager/argocd/extract.ts [hunk 5]
```diff
@@ -88,15 +107,31 @@ function processAppSpec(
...
+
+  const parts = kustomizeImageRe.exec(image);
+  if (parts?.length !== 3) {
```
[the remark this answers is not in the document]
Comment (j-be) [reply lib/modules/manager/argocd/extract.ts]: Fair enough, changed.

File: lib/modules/manager/argocd/extract.ts [hunk 6]
```diff
@@ -88,15 +107,24 @@ function processAppSpec(
...
   for (const source of coerceArray(spec.sources)) {
-    deps.push(processSource(source));
+    deps = deps.concat(processSource(source));
+  }
+
+  return deps;
+}
+
+function processKustomizeImage(image: string): PackageDependency | null {
+  const parts = kustomizeImageRe.exec(image);
```
Comment (j-be) [inline lib/modules/manager/argocd/extract.ts:124]: I'm not quite happy with having both the parameter to the function, as well as the capture group named `image`, even though both are accurate, as the former is an item from `images` list, the latter is - well - an image.

Do you have a suggestion for a better name of either of the two?

File: lib/modules/manager/argocd/__fixtures__/validApplication.yml [hunk 7]
```diff
@@ -128,6 +133,7 @@ spec:
       helm:
         valueFiles:
           - $foo/values.yaml
+
```
Comment (viceice) [inline lib/modules/manager/argocd/__fixtures__/validApplication.yml:136]: revert

File: lib/modules/manager/argocd/extract.ts [hunk 8]
```diff
@@ -59,26 +63,41 @@ function processSource(source: ApplicationSource): PackageDependency | null {
...
       currentValue: source.targetRevision,
-      datasource: HelmDatasource.id,
-    };
+      datasource: GitTagsDatasource.id,
+    },
+  ];
+
+  // Git repo is pointing to a Kustomize resources
+  if (source.kustomize?.images) {
+    dependencies = dependencies.concat(
```
Comment (viceice) [inline lib/modules/manager/argocd/extract.ts:95]:
```suggestion
    dependencies.push(
```

File: lib/modules/manager/argocd/extract.ts [hunk 9]
```diff
@@ -59,26 +63,41 @@ function processSource(source: ApplicationSource): PackageDependency | null {
...
+      source.kustomize.images.map(processKustomizeImage).filter(is.truthy),
```
Comment (viceice) [inline lib/modules/manager/argocd/extract.ts:96]:
```suggestion
      ...source.kustomize.images.map(processKustomizeImage).filter(is.truthy),
```

File: lib/modules/manager/argocd/extract.ts [hunk 10]
```diff
@@ -89,15 +108,24 @@ function processAppSpec(
       ? definition.spec
       : definition.spec.template.spec;
 
-  const deps: (PackageDependency | null)[] = [];
+  let deps: PackageDependency[] = [];
 
   if (is.nonEmptyObject(spec.source)) {
-    deps.push(processSource(spec.source));
+    deps = processSource(spec.source);
```
Comment (viceice) [inline lib/modules/manager/argocd/extract.ts:114]: see above

File: lib/modules/manager/argocd/extract.ts [hunk 11]
```diff
@@ -89,15 +108,24 @@ function processAppSpec(
...
   }
 
   for (const source of coerceArray(spec.sources)) {
-    deps.push(processSource(source));
+    deps = deps.concat(processSource(source));
```
Comment (viceice) [inline lib/modules/manager/argocd/extract.ts:118]: same

File: lib/modules/manager/argocd/extract.ts [hunk 12]
```diff
@@ -89,15 +108,24 @@ function processAppSpec(
...
+  }
+
+  return deps;
+}
+
+function processKustomizeImage(image: string): PackageDependency | null {
```
Comment (viceice) [inline lib/modules/manager/argocd/extract.ts:124]:
```suggestion
function processKustomizeImage(kustomizeImage: string): PackageDependency | null {
```

File: lib/modules/manager/argocd/extract.ts (hunk 6 above)
Comment (viceice) [reply lib/modules/manager/argocd/extract.ts:125]: see above

File: lib/modules/manager/argocd/__fixtures__/validApplication.yml (hunk 7 above)
Comment (j-be) [reply lib/modules/manager/argocd/__fixtures__/validApplication.yml]: Done

File: lib/modules/manager/argocd/extract.ts (hunk 8 above)
Comment (j-be) [reply lib/modules/manager/argocd/extract.ts]: Changed

File: lib/modules/manager/argocd/extract.ts (hunk 9 above)
Comment (j-be) [reply lib/modules/manager/argocd/extract.ts]: Changed

File: lib/modules/manager/argocd/extract.ts (hunk 10 above)
Comment (j-be) [reply lib/modules/manager/argocd/extract.ts]: Changed

File: lib/modules/manager/argocd/extract.ts (hunk 11 above)
Comment (j-be) [reply lib/modules/manager/argocd/extract.ts]: Changed

File: lib/modules/manager/argocd/extract.ts (hunk 12 above)
Comment (j-be) [reply lib/modules/manager/argocd/extract.ts]: Changed

File: lib/modules/manager/argocd/extract.ts [hunk 13]
```diff
@@ -18,6 +20,8 @@ import {
 } from './schema';
 import { fileTestRegex } from './util';
 
+const kustomizeImageRe = regEx(/.+=(?<image>.+)/);
```
Comment (viceice) [inline lib/modules/manager/argocd/extract.ts:23]:
```suggestion
const kustomizeImageRe = regEx(/=(?<image>.+)$/);
```

**Review by viceice — DISMISSED** [review]

**Review by secustor — COMMENTED** [review]
@j-be  
This needs some additional test cases for coverage.

Do you plan to continue the work on this?

**Comment by j-be** [comment]
@secustor I do. What is missing? I Thought I adtessed all comments.

File: lib/modules/manager/argocd/extract.ts [hunk 14]
```diff
@@ -49,33 +53,48 @@ export function extractPackageFile(
...
+      depName: source.repoURL,
       currentValue: source.targetRevision,
-      datasource: HelmDatasource.id,
-    };
+      datasource: GitTagsDatasource.id,
+    },
+  ];
+
+  // Git repo is pointing to a Kustomize resources
+  if (source.kustomize?.images) {
```
Comment (viceice) [inline lib/modules/manager/argocd/extract.ts:91]: kustomize is only allowed with git repos? not allowed with chart?

Comment (j-be) [reply lib/modules/manager/argocd/extract.ts:91]: Not sure I understand: why and how would I use Kustomize from within a Helm Chart?

**Comment by j-be** [comment]
@secustor Just noticed it shows in the diff of the PR. I see what is missing. Will add shortly.

Sorry for that.

File: lib/modules/manager/argocd/extract.ts (hunk 14 above)
Comment (viceice) [reply lib/modules/manager/argocd/extract.ts:91]: I don't know, I'm not an argocd user

Comment (j-be) [reply lib/modules/manager/argocd/extract.ts:91]: For all I know: Yes, image overrides for Kustomize only make sense with Git, not with Chart.

**Review by viceice — APPROVED** [review]

**Review by secustor — APPROVED** [review]

**Comment by secustor** [comment]
Thanks for the contribution!

**Comment by j-be** [comment]
You're welcome 😄

**Pull request merged**